pkg/apimachinery/patch/patch.go

```diff
@@ -38,9 +37,73 @@ const (
 	PatchRemoveOp  = "remove"
 )
 
+type PatchSet interface {
+	Test(path string, value interface{})
+	Add(path string, value interface{})
+	Remove(path string)
+	Replace(path string, value interface{})
+	TestAndReplace(path string, oldVal, newVal interface{})
+	AddOrReplace(path string, value interface{}, condition bool)
+	IsEmpty() bool
+	GeneratePayload() ([]byte, error)
+}
+
+type patchSet struct {
+	patches []PatchOperation
+}
+
+func New() PatchSet {
+	return &patchSet{}
+}
+
+func (p *patchSet) IsEmpty() bool {
+	return len(p.patches) < 1
+}
+
+func (p *patchSet) addOp(op, path string, value interface{}) {
+	p.patches = append(p.patches, PatchOperation{
+		Op:    op,
+		Path:  path,
+		Value: value,
+	})
+}
+
+func (p *patchSet) Test(path string, value interface{}) {
+	p.addOp(PatchTestOp, path, value)
+}
+
+func (p *patchSet) Add(path string, value interface{}) {
+	p.addOp(PatchAddOp, path, value)
+}
+
+func (p *patchSet) Remove(path string) {
+	p.addOp(PatchRemoveOp, path, nil)
+}
+
+func (p *patchSet) Replace(path string, value interface{}) {
+	p.addOp(PatchReplaceOp, path, value)
+}
+
+func (p *patchSet) TestAndReplace(path string, oldVal, newVal interface{}) {
```

Comment:
Similar to `AddOrReplace`, if one does a `test`, I find it hard to see a use case that needs`replace`. An `add` will work better in most cases.

But I would like to propose that this object will not have all these combinations, it should just support the basic commands.

A factory can be added to build a collections of operations, with a build pattern. Then, whatever combination is required, it can be built and avoid having such a matrix of methods.

Example:
```go
patchSet := patch.New(
  patch.WithAdd(...)
  patch.WithTest(...)
)

vm, err := kubevirt.Client().VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, patchSet.Payload(), &metav1.PatchOptions{}
```

Reply:
@EdDev see for example this discussion: https://github.com/kubevirt/kubevirt/pull/11211#discussion_r1502554835

Reply:
The example you linked is good. The code there is wrong and redundant.

```go
if vm.Status.MemoryDumpRequest == nil {
	patches.Add("/status/memoryDumpRequest", vmCopy.Status.MemoryDumpRequest)
} else {
	patches.Replace("/status/memoryDumpRequest", vmCopy.Status.MemoryDumpRequest)
```

Is equivalent to a simple:
```go
patches.Add("/status/memoryDumpRequest", vmCopy.Status.MemoryDumpRequest)
```

`add` overrides the existence, same as `replace`.
The only difference between the two is that `replace` will blow up if the path was not there before.

Reply:
Are we sure? Checking the [documentation](https://jsonpatch.com/#operations), it says `Replaces a value. Equivalent to a “remove” followed by an “add”.`

Reply:
I thought `add` should fail if the value already exists and not replacing it. I'll give a try with a small example

Reply:
> Are we sure? Checking the [documentation](https://jsonpatch.com/#operations), it says `Replaces a value. Equivalent to a “remove” followed by an “add”.`

Yes, `remove` will fail if the given path is not there and that is essentially the main different between the two.

Reply:
Ok, thanks! Make sense and we can further simplify the code. I just tried to stick to what we had in the previous version of the code.

I would also need to heavily change the unit tests as basically the check the same sequence of patches we have in the flow.

Reply:
IMO it will help if you first create the initial object and use it to cover all its API. Then, in follow up PRs replace to use it.

Trying to do everything in one PR is too challenging and it will be hard to merge without rebasing all the time.

Reply:
Not sure, I understand. Do you meant just to add the new type in the code with the unit tests and then in a follow-up PR replacing the existing operations?

Reply:
@EdDev @0xFelix something like: https://github.com/kubevirt/kubevirt/pull/11599 ? We could initially replace all the `GeneratePatchPayload` and then in a following PR all the hardcoded operations as I did in this PR. WDYT?

Reply:
My suggestion was to create the new object with all its methods and its unit tests.
Continue in other commits to replace places in the codebase to use it... at least one time each method.
This should give the reviewers the chance to see how it is used and convince them it makes sense.

Then, in follow up PRs to start replacing places in the codebase one by one (or as a group).